lib/Server.js

```diff
@@ -3255,9 +3255,70 @@ class Server {
    * @param {string | string[]} watchPath
    * @param {WatchOptions} [watchOptions]
    */
-  watchFiles(watchPath, watchOptions) {
+  watchFiles(watchPath, watchOptions = {}) {
     const chokidar = require("chokidar");
-    const watcher = chokidar.watch(watchPath, watchOptions);
+
+    const watchPathArr = Array.isArray(watchPath) ? watchPath : [watchPath];
+
+    if (watchOptions.disableGlobbing !== true) {
+      const isGlob = require("is-glob");
+      const watchPathGlobs = watchPathArr.filter((p) => isGlob(p));
+
+      // No need to do all this work when no globs are used
+      if (watchPathGlobs.length > 0) {
+        const globParent = require("glob-parent");
+        const picomatch = require("picomatch");
+
+        watchPathGlobs.forEach((p) => {
+          watchPathArr[watchPathArr.indexOf(p)] = globParent(p);
+        });
+
+        const matcher = picomatch(watchPathGlobs, {
+          cwd: watchOptions.cwd,
+          dot: true,
+        });
+        const ignoreFunc = (/** @type {string} */ p) =>
+          !watchPathArr.includes(p) && !matcher(p);
+
+        if (Array.isArray(watchOptions.ignored)) {
+          const ignoredGlobs = [];
+          for (let i = 0; i < watchOptions.ignored.length; i++) {
+            const ignored = watchOptions.ignored[i];
+            if (typeof ignored === "string" && isGlob(ignored)) {
+              ignoredGlobs.push(ignored);
+              watchOptions.ignored.splice(i, 1);
+            }
+          }
+
+          if (ignoredGlobs.length > 0) {
+            const ignoreMatcher = picomatch(ignoredGlobs, {
+              dot: true,
+              cwd: watchOptions.cwd,
+            });
+            watchOptions.ignored.push(ignoreMatcher);
+          }
+
+          watchOptions.ignored.push(ignoreFunc);
+        } else {
+          if (
+            watchOptions.ignored &&
+            typeof watchOptions.ignored === "string" &&
+            isGlob(watchOptions.ignored)
+          ) {
+            watchOptions.ignored = picomatch(watchOptions.ignored, {
+              dot: true,
+              cwd: watchOptions.cwd,
+            });
+          }
```

Comment:
Can we simplify it because there is a lot of duplicate logic here? We can always convert `watchOptions.ignored` to array (when it is provided)

Reply:
agreed! will do